xds/src/main/java/io/grpc/xds/AddressFiltering.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.Attributes;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.NameResolver.ResolutionResultAttr;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+final class AddressFiltering {
+  @ResolutionResultAttr
+  private static final Attributes.Key<PathChain> PATH_CHAIN_KEY =
+      Attributes.Key.create("io.grpc.xds.AddressFilter.PATH_CHAIN_KEY");
+
+  // Prevent instantiation.
+  private AddressFiltering() {}
+
+  /**
+   * Sets a path filter to an EquivalentAddressGroup. Any preexisting path filter will be discarded.
+   */
+  static EquivalentAddressGroup setPathFilter(EquivalentAddressGroup address, List<String> names) {
+    checkNotNull(address, "address");
+    checkNotNull(names, "names");
+    Attributes.Builder attrBuilder = address.getAttributes().toBuilder().discard(PATH_CHAIN_KEY);
+    PathChain pathChain = null;
```

Comment:
A simple `List` or `Queue` works well,  I don't see the necessity of implementing such a "linked nodes".

Reply:
For a `List`, you either have to maintain an index, or make a copy sublist each time when you filter to the next hierarchy. As for `Queue`, the `pop()` method may mutate the original address. 

The linked node is a private class hinden from the user.

Reply:
`LinkedList` works perfectly.

> As for Queue, the pop() method may mutate the original address.

This argument does not make sense. Everything happens internally inside the utility method's implementation, the only person that mutates the address is the author of this class. There is no information leaking.

Reply:
> This argument does not make sense. Everything happens internally inside the utility method's implementation, the only person that mutates the address is the author of this class. There is no information leaking.

There is no point to use `pop()` for the author of this class. You still need to make a copy, because the original address should not be mutated when you call `filter()`.

Reply:
I guess you really mean `poll()` or `remove()`.

Why the address (except its attribute) will be mutated? Each `EquivalentAddressGroup` holds a `LinkedList` and each time you do filter, just peeks/polls the first element from the list and done. With current implementation, you holding the head of linked nodes, and each time you "poll" the head, you need to update the reference to point to the next node.

Reply:
> Why the address (except its attribute) will be mutated?

I mean the attribute of the original address will be mutated. We should not.

I don't mind to make a copy sublist each time calling `filter()` if you prefer. In that case, an ArrayList will do.

Reply:
> I mean the attribute of the original address will be mutated. We should not.

Why is that related to what we are discussing here? You do the same thing as what you are doing now:  
`return new EquivalentAddressGroup(address.getAddresses(), attrBuilder.build());`

The only change is `PATH_CHAIN_KEY` is a key for a `List` (implementation-wise, a `LinkedList`).

> I don't mind to make a copy sublist each time calling filter() if you prefer.

Why calling `filter()` each time will make a copy? You just poll the first element from the list, that's an O(1) operation for `LinkedList`.

Reply:
>> I mean the attribute of the original address will be mutated. We should not.

> Why is that related to what we are discussing here? You do the same thing as what you are doing now:
return new EquivalentAddressGroup(address.getAddresses(), attrBuilder.build());

That's totally different (I'm not talking about `setPathFilter()`). The current implementation of `filter()` doesn't mutated the input EAGs or their attributes. The `poll()` will, and that's not allowed.

Reply:
> The current implementation of filter() doesn't mutate the input EAGs or their attributes. 

No, that should not be the case. You should not always return a new `EquivalentAddressGroup`. We are using `PATH_CHAIN_KEY` as an internal trick to mark and pass address information. This attribute is only known to those two layers of LB policies. Keeping input `EquivalentAddressGroup`s of `filter()` is a redundant requirement.

Reply:
>Keeping input EquivalentAddressGroups of filter() is a redundant requirement.

No, it is a requirement. You must not mutate the input. For example, it's possible that the input EAGs are cached and used for the child LBs next time, in which case `poll()` twice is illegal.

Reply:
Ok, that makes sense. So these two utility methods should really be free of side effect. Then you might want to update the comments and method names, current ones are misleading.